### DISCUSSION TITLE
Built-in unnecessary rendering management with Remix ?

### DISCUSSION THREAD

**Discussion by kael** [remix-run/remix#4182] [Q&A] [answered]
I'm writing a complex WebApp and getting overwhelmed with lots of unnecessary cascade rendering.

Learning to tame `useCallback`, `useMemo`, `memo` and `useEffect`, there seems to be systematic requirement of a boilerplate that can't be avoid for each component so that it only renders when really necessary.

Does Remix auto-magically handle some auto-memoization or prevent unnecessary rendering ? Does it use singleton-like data hooks which won't be reinitialized each time the component renders unnecessarily ?

**Comment by sergiodxa** [comment, marked as answer]
The data hooks (useLoaderData and useActionData) have their data stored in an internal single context object so they will trigger a re-render only when that change, also every hook uses useCallback and useMemo internally to ensure your UI is not re-rendering if not needed.

**Reply by kael** [reply]
Thank you very much for this answer. :+1: 

I'm going to give a try to Remix.
